pystoned/CNLS.py

```diff
@@ -1,304 +1,224 @@
-                               sum(model.b[h, j] * arow[j] for j in model.j)
-
-                    model.qconcav = Constraint(model.i, model.h, rule=qconcav_rule, doc='concavity constraint')
-
-                # cost model
-                if fun == "cost":
-
-                    def qconvex_rule(model, i, h):
-                        arow = x[i]
-                        if i == h:
-                            return Constraint.Skip
-                        return sum(model.b[i, j] * arow[j] for j in model.j) >= \
-                               sum(model.b[h, j] * arow[j] for j in model.j)
-
-                    model.qconvex = Constraint(model.i, model.h, rule=qconvex_rule, doc='convexity constraint')
-
-    return model
\ No newline at end of file
+# Import pyomp module
+from pyomo.environ import *
+from pyomo.core.expr.numvalue import NumericValue
+
+
+class CNLS:
+    def __init__(self, y, x, cet='addi', fun='prod', rts='vrs'):
+        # cet     = "addi" : Additive composite error term
+        #         = "mult" : Multiplicative composite error term
+        # fun     = "prod" : production frontier
+        #         = "cost" : cost frontier
+        # rts     = "vrs"  : variable returns to scale
+        #         = "crs"  : constant returns to scale
+
+        ## TODO(error/warning handling): Check the configuration of the model exist
+        self.x = x.tolist()
+        self.y = y.tolist()
+        self.cet = cet
+        self.fun = fun
+        self.rts = rts
+
+        if type(self.x[0]) != list:
+            self.x = [x.tolist()]
+
+        # Initialize the pyomo model
+        self.__model__ = ConcreteModel()
+        self.__model__.I = Set(initialize=range(len(self.y)))
+        self.__model__.J = Set(initialize=range(len(self.x[0])))
+
+        # Initialize the variables
+        self.__model__.alpha = Var(self.__model__.I, doc='alpha')
+        self.__model__.beta = Var(
+            self.__model__.I, self.__model__.J, bounds=(0.0, None), doc='beta')
+        self.__model__.epsilon = Var(self.__model__.I, doc='residuals')
+        self.__model__.frontier = Var(
+            self.__model__.I, bounds=(0.0, None), doc='estimated frontier')
+
+        # Setup the objective function and constraints for model
+        self.__model__.objective = Objective(
+            rule=self.__objective_rule(),
+            sense=minimize,
+            doc='objective function')
+        self.__model__.regression_rule = Constraint(
+            self.__model__.I,
+            rule=self.__regression_rule(),
+            doc='regression equation')
+        if self.cet == "mult":
+            self.__model__.log_rule = Constraint(
+                self.__model__.I,
+                rule=self.__log_rule(),
+                doc="regression function")
+        self.__model__.concav_convex_rule = Constraint(
+            self.__model__.I,
+            self.__model__.I,
+            rule=self.__concav_convex_rule(),
```

Comment:
The `__concav_convex_rule()` is a little bit longer, so we can replace it with `__afriat_rule()` here.

Reply:
I see. I tried to figure out which word to describe too.
Thanks for your recommendation!

Reply:
OK.